pep-9999.rst

```diff
@@ -0,0 +1,675 @@
+            ) -> IO[bytes]: ...
+
+   # Fallback overload for when the user isn't using literal types
+   @overload
+   def open(path: _PathType, mode: str) -> IO[Any]: ...
+
+If we change the signature of ``open`` to use just the first two overloads,
+we would break any code that does not pass in a literal string expression.
+For example, code like this would be broken::
+
+   mode: str = pick_file_mode(...)
+   with open(path, mode) as f:
+       # f should continue to be of type IO[Any] here
+
+A little more broadly: we propose adding a policy to typeshed that
+mandates that whenever we add literal types to some existing API, we also
+always include a fallback overload to maintain backwards-compatibility.
+
+Interactions with generics
+''''''''''''''''''''''''''
+
+Types like ``Literal[3]`` are meant to be just plain old subclasses of
+``int``. This means you can use types like ``Literal[3]`` anywhere
+you could use normal types, such as with generics.
+
+For example, suppose we want to construct a type representing a
+2-dimensional Matrix which can be parameterized by two literal ints
+representing the number of rows and columns respectively. Such a type
+could be built using the existing generics system like so::
+
+   A = TypeVar('A', bound=int)
+   B = TypeVar('B', bound=int)
+   C = TypeVar('C', bound=int)
+
+   # A simplified definition for Matrix[row, column]
+   class Matrix(Generic[A, B]):
+       def __init__(self, elements: List[List[int]]) -> None: ...
+       def __add__(self, other: Matrix[A, B]) -> Matrix[A, B]: ...
+       def __matmul__(self, other: Matrix[B, C]) -> Matrix[A, C]: ...
+       def transpose(self) -> Matrix[B, A]: ...
+       
+   Foo: Matrix[Literal[2], Literal[3]] = Matrix(...)
+   Bar: Matrix[Literal[3], Literal[7]] = Matrix(...)
+
+   reveal_type(Foo @ Bar)  # Revealed type is Matrix[Literal[2], Literal[7]]
+   Bar @ Foo               # Error, Foo doesn't match expected type Matrix[Literal[7], Literal[int]]
+
+This class definition is not perfect: it would not prohibit users from
+constructing less precise types like ``Matrix[int, int]`` due to the
+typevar bound, for example.
+
+We considered several different proposals for addressing this gap 
+but ultimately rejected all of them and decided to defer the problem
+of integer generics to a later date. See `Rejected or out-of-scope ideas`_
+
+Interactions with asserts and other checks
+''''''''''''''''''''''''''''''''''''''''''
+
+Type checkers should narrow the type of variables when they are compared
+directly against other literal types. For example::
+
+   def foo(x: str) -> None:
+       if x == "foo":
+           # Type checker should narrow 'x' to "foo" here
+           expects_foo(x)
+       
+       # Similarly, type checker should narrow 'x' to "bar" here
+       assert x == "bar"
+       expects_bar(x)
+
+Type checkers may optionally perform additional analysis and narrowing.
```

Comment:
We need to say something about possibility of warnings in case of non-exhaustive checks. This was requested by some users.

Reply:
I mentioned this in the email from earlier, but I think talking about warnings in the case of non-exhaustive checks really has more to do with the semantics of Unions more then Literal types.

I did add a note reminding people of the [section in PEP 484 that discussions enums](https://www.python.org/dev/peps/pep-0484/#support-for-singleton-types-in-unions) and implied that some type checkers could perform reachability/exhaustiveness checks in the same kind of way we imply that in the section about [NoReturn in PEP 484](https://www.python.org/dev/peps/pep-0484/#the-noreturn-type).

I think nailing to what degree type checkers should perform non-exhaustive checks vs requiring users to use tricks like `assert_never(...)`, to what degree type checkers should detect and warn about dead code, etc should probably be an entirely separate discussion.